include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -360,19 +362,20 @@ struct __parallel_transform_reduce_impl
                         auto __group_idx = __item_id.get_group(0);
                         // 1. Initialization (transform part). Fill local memory
                         _Size __n_items;
+
+                        _Tp __result = *std::launder(reinterpret_cast<_Tp*>(alloca(sizeof(_Tp))));
```

Comment:
Can we come up with a simpler approach and one that doesn't potentially lead to a stack overflow?

Reply:
This is something that is on the more risky side of this PR. I am using this approach to get around having no guarantees of `_Tp` having a default constructor. Essentially we are just allocating garbage memory on the stack. This should be safe so long as `_Tp` is not too large.  I believe we are already in a similar position as if `_Tp` is too large we cannot allocate enough local memory to accommodate the data type.

So far as I am aware this is the most suitable approach that does not use local memory, to get around this issue. Other approaches would require a default constructor.

The other alternative that I can think of might be do allocate an array of `char`s of equal size to `_Tp` then reinterpret cast this as the return value. I believe these two approaches would be essentially equivalent, however if you have a preference for the latter I can explore that option.

Reply:
Could you update the PR to include your explanation as a comment? I think it would be helpful to understand the usage of `alloca` here.

Reply:
Other algorithms in oneDPL have used unions to support types that are not default constructible. For example, radix sort: https://github.com/oneapi-src/oneDPL/blob/516984dbc8808f1a369816005af642ddfc6fb594/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h#L167

Could something like that be used here?

Reply:
Thanks for the suggestion. I have moved the code over to use them instead of alloca